test/e2e/operator.go

```diff
@@ -272,3 +276,108 @@ var _ = Describe("ARO Operator - Conditions", func() {
+		dummynsg = mgmtnetwork.SecurityGroup{
+			Location:                      to.StringPtr(location),
+			Name:                          to.StringPtr("dummynsg"),
+			Type:                          to.StringPtr("Microsoft.Network/networkSecurityGroups"),
+			SecurityGroupPropertiesFormat: &mgmtnetwork.SecurityGroupPropertiesFormat{},
+		}
+		err := clients.NetworkSecurityGroups.CreateOrUpdateAndWait(ctx, resourceGroup, "dummynsg", dummynsg)
+		Expect(err).NotTo(HaveOccurred())
+		dummynsgObject, err := clients.NetworkSecurityGroups.Get(ctx, resourceGroup, "dummynsg", "")
+		Expect(err).NotTo(HaveOccurred())
+		dummynsg.ID = dummynsgObject.ID
+	}
+	BeforeEach(func() {
+		enableReconcileNSG()
+		gatherNetworkInfo()
+		createDummyNSG()
+	})
+	AfterEach(func() {
+		err := clients.NetworkSecurityGroups.DeleteAndWait(context.Background(), resourceGroup, "dummynsg")
+		if err != nil {
+			log.Warn(err)
+		}
+	})
+	Specify("ARO Operator - Azure NSG Reconciler must reconcile worker subnet when NSG is changed", func() {
+		wsubnetObject, err := clients.Subnet.Get(ctx, resourceGroup, vnetName, workerSubnet, "")
+		Expect(err).NotTo(HaveOccurred())
+		correctWorkerNSG := wsubnetObject.NetworkSecurityGroup.ID
+
+		wsubnetObject.NetworkSecurityGroup = &dummynsg
+		err = clients.Subnet.CreateOrUpdateAndWait(ctx, resourceGroup, vnetName, workerSubnet, wsubnetObject)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = wait.PollImmediate(30*time.Second, 10*time.Minute, func() (bool, error) {
+			ws, err := clients.Subnet.Get(ctx, resourceGroup, vnetName, workerSubnet, "")
+			if err != nil {
+				return false, err
+			}
+			if *ws.NetworkSecurityGroup.ID == *correctWorkerNSG {
+				log.Info("worker subnetnsg matched expected value")
+				return true, nil
+			}
+			log.Errorf("worker nsg: %s did not match expected value: %s", *ws.NetworkSecurityGroup.ID, *correctWorkerNSG)
+			return false, nil
+		})
+		Expect(err).NotTo(HaveOccurred())
+	})
+	Specify("ARO Operator - Azure NSG Reconciler must reconcile master subnet when NSG is changed", func() {
+		msubnetObject, err := clients.Subnet.Get(ctx, resourceGroup, vnetName, masterSubnet, "")
+		Expect(err).NotTo(HaveOccurred())
+		correctMasterNSG := msubnetObject.NetworkSecurityGroup.ID
+
+		msubnetObject.NetworkSecurityGroup = &dummynsg
+		err = clients.Subnet.CreateOrUpdateAndWait(ctx, resourceGroup, vnetName, masterSubnet, msubnetObject)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = wait.PollImmediate(30*time.Second, 10*time.Minute, func() (bool, error) {
+			ms, err := clients.Subnet.Get(ctx, resourceGroup, vnetName, masterSubnet, "")
+			if err != nil {
+				return false, err
+			}
+			if *ms.NetworkSecurityGroup.ID == *correctMasterNSG {
+				log.Info("master subnetnsg matched expected value")
+				return true, nil
+			}
+			log.Errorf("master nsg: %s did not match expected value: %s", *ms.NetworkSecurityGroup.ID, *correctMasterNSG)
+			return false, nil
+		})
+		Expect(err).NotTo(HaveOccurred())
+	})
+})
```

Comment:
These two tests look identical to me. The only difference I is master/worker subnet. Can create a function which accepts params so we can resue it in these two tests?

Reply:
I just decided to loop over steps for each subnet we need to test instead of a func call.